pkg/operation/botanist/dns/dnsprovider.go

```diff
@@ -0,0 +1,137 @@
+	Name       string            `json:"name,omitempty"`
+	Purpose    string            `json:"purpose,omitempty"`
+	Provider   string            `json:"provider,omitempty"`
+	Labels     map[string]string `json:"labels,omitempty"`
+	SecretData map[string][]byte `json:"secretData,omitempty"`
+	Domains    *IncludeExclude   `json:"domains,omitempty"`
+	Zones      *IncludeExclude   `json:"zones,omitempty"`
+}
+
+// IncludeExclude contain slices of excluded and included domains/zones.
+type IncludeExclude struct {
+	Include []string `json:"include,omitempty"`
+	Exclude []string `json:"exclude,omitempty"`
+}
+
+// NewDNSProvider creates a new instance of DeployWaiter for a specific DNS provider.
+// <waiter> is optional and it's defaulted to github.com/gardener/gardener/pkg/utils/retry.DefaultOps()
+func NewDNSProvider(
+	values *ProviderValues,
+	shootNamespace string,
+	applier kubernetes.ChartApplier,
+	chartsRootPath string,
+	logger *logrus.Entry,
+	client crclient.Client,
+	waiter retry.Ops,
+) component.DeployWaiter {
+	if waiter == nil {
+		waiter = retry.DefaultOps()
+	}
+
+	return &dnsProvider{
+		values:         values,
+		shootNamespace: shootNamespace,
+		ChartApplier:   applier,
+		chartPath:      filepath.Join(chartsRootPath, "seed-dns", "provider"),
+		logger:         logger,
+		client:         client,
+		waiter:         waiter,
+	}
+}
+
+type dnsProvider struct {
+	values         *ProviderValues
+	shootNamespace string
+	kubernetes.ChartApplier
+	chartPath string
+	logger    *logrus.Entry
+	client    crclient.Client
+	waiter    retry.Ops
+}
+
+func (d *dnsProvider) Deploy(ctx context.Context) error {
+	return d.Apply(ctx, d.chartPath, d.shootNamespace, d.values.Name, kubernetes.Values(d.values))
+}
+
+func (d *dnsProvider) Destroy(ctx context.Context) error {
+	return client.IgnoreNotFound(d.client.Delete(ctx, d.provider()))
+}
+
+func (d *dnsProvider) Wait(ctx context.Context) error {
+	var (
+		status  string
+		message string
+	)
+
+	if err := d.waiter.UntilTimeout(ctx, 5*time.Second, 2*time.Minute, func(ctx context.Context) (done bool, err error) {
+		provider := &dnsv1alpha1.DNSProvider{}
+		if err := d.client.Get(
+			ctx,
+			crclient.ObjectKey{Name: d.values.Name, Namespace: d.shootNamespace},
+			provider,
+		); err != nil {
+			return retry.SevereError(err)
+		}
+
+		if provider.Status.State == dnsv1alpha1.STATE_READY {
+			return retry.Ok()
+		}
+
+		status = provider.Status.State
+		if msg := provider.Status.Message; msg != nil {
+			message = *msg
+		}
+
+		d.logger.Infof("Waiting for %q DNS provider to be ready... (status=%s, message=%s)", d.values.Name, status, message)
+		return retry.MinorError(fmt.Errorf("DNS provider %q is not ready (status=%s, message=%s)", d.values.Name, status, message))
+	}); err != nil {
+		return gardencorev1beta1helper.DetermineError(err, fmt.Sprintf("Failed to create DNS provider for %q DNS record: %q (status=%s, message=%s)", d.values.Name, err.Error(), status, message))
```

Comment:
FYI: [This change](https://github.com/gardener/gardener/pull/2336/files#diff-33d2a8cacfcd2dda84699a11862f3ab0) should be reflected here as well as #2336 will probably be merged first.